apps/src/blockly/addons/cdoFieldImageDropdown.js

```diff
@@ -20,20 +20,94 @@ export class CdoFieldImageDropdown extends FieldGridDropdown {
+   */
+  showEditor_(opt_e = undefined) {
+    this.dropdownCreate_();
+    if (opt_e && typeof opt_e.clientX === 'number') {
+      this.menu_.openingCoords = new Blockly.utils.Coordinate(
+        opt_e.clientX,
+        opt_e.clientY
+      );
+    } else {
+      this.menu_.openingCoords = null;
+    }
+
+    if (this.buttons_) {
+      // Force buttons to a new row by adding blank elements if needed.
+      const numItems = this.menu_.menuItems_.length;
+      const numInLastRow = numItems % this.columns_;
+      const numBlankToAdd = numInLastRow > 0 ? this.columns_ - numInLastRow : 0;
+      for (let i = 0; i < numBlankToAdd; i++) {
+        const item = document.createElement('div');
+        item.style.width = this.imageWidth_ + 'px';
+        item.style.height = this.imageHeight_ + 'px';
+        const menuItem = new Blockly.MenuItem(item, '');
+        menuItem.setEnabled(false);
+        this.menu_.addChild(menuItem);
+      }
+
+      // Add buttons to menu
+      this.buttons_.forEach(button => {
+        const buttonElement = document.createElement('BUTTON');
+        buttonElement.innerHTML = button.text;
+        buttonElement.addEventListener('click', button.action);
+        buttonElement.addEventListener('click', () =>
+          Blockly.DropDownDiv.hideIfOwner(this, true)
+        );
+        const menuItem = new Blockly.MenuItem(buttonElement, '');
+        menuItem.setRole(Blockly.utils.aria.Role.OPTION);
+        menuItem.setRightToLeft(this.sourceBlock_.RTL);
+        menuItem.setEnabled(false);
+        this.menu_.addChild(menuItem);
+      });
+    }
+
+    // Element gets created in render.
+    this.menu_.render(Blockly.DropDownDiv.getContentDiv());
+    var menuElement = /** @type {!Element} */ (this.menu_.getElement());
+    Blockly.utils.dom.addClass(menuElement, 'blocklyDropdownMenu');
 
-    // Override so that grid dropdown is white.
-    // The Blockly team is planning to update the FieldGridDropdown plugin
-    // so that the dropdown color is configurable. Once that work is done, we can
-    // remove this code.
     const primaryColour = color.white;
     const borderColour = this.sourceBlock_.isShadow()
       ? this.sourceBlock_.getParent().style.colourTertiary
       : this.sourceBlock_.style.colourTertiary;
     Blockly.DropDownDiv.setColour(primaryColour, borderColour);
+
+    // Focusing needs to be handled after the menu is rendered and positioned.
+    // Otherwise it will cause a page scroll to get the misplaced menu in
+    // view. See issue #1329.
+    this.menu_.focus();
+
+    if (this.selectedMenuItem_) {
+      this.menu_.setHighlighted(this.selectedMenuItem_);
+    }
+
+    this.applyColour();
+
+    Blockly.utils.dom.addClass(
+      this.menu_.getElement(),
+      'fieldGridDropDownContainer'
+    );
+    this.updateColumnsStyling_();
+
+    Blockly.DropDownDiv.showPositionedByField(
+      this,
+      this.dropdownDispose_.bind(this)
+    );
```

Comment:
i think i understand what this function does ([this](https://github.com/google/blockly/blob/64188ae27297b1ca23a3244cc987dfbfc6e98f34/core/dropdowndiv.js#L266-L284) is the implementation i'm looking at), but i don't really understand why we need to call it here

Reply:
This is what actually opens the dropdown. All dropdowns in Blockly are actually just one div that moves around, so this method gets called by the field that wants to open a dropdown and it positions the DropDownDiv adjacent to that field. So if we don't call this method in `FieldImageDropdown.showEditor_`, the dropdown won't actually show.

Reply:
ahh, okay i definitely didn't understand what that function does then 😄